src/main/groovy/io/openliberty/tools/gradle/tasks/AbstractServerTask.groovy

```diff
@@ -731,21 +745,103 @@ abstract class AbstractServerTask extends AbstractTask {
         }
     }
 
-    private void writeServerEnvProperties(File file, Properties properties, Properties projectProperties) throws IOException {
-        Properties combinedEnvProperties = null
+    private void handleServerEnvFileAndProperties(String serverEnvPath, String serverDirectory) {
+        File envFile = new File(serverDirectory, "server.env")
+        Properties configuredProps = combineServerEnvProperties(server.env, envProjectProps);
+
+        if(server.appendServerEnv) {
+            setServerEnvWithAppendServerEnvHelper(envFile, serverEnvPath, configuredProps)
+        }
+        else {
+            setServerEnvHelper(envFile, serverEnvPath, configuredProps)
+        }
+    }
+
+    private void setServerEnvWithAppendServerEnvHelper(File envFile, String serverEnvPath, Properties configuredProps) {
+        Properties serverEnvProps = convertServerEnvToProperties(envFile);
+        Properties mergedProperties = new Properties();
+
+        if (server.serverEnvFile != null && server.serverEnvFile.exists()) {
+            if (serverEnvPath != null) {
+                logger.warn("The " + serverEnvPath + " file is merged with the " + server.serverEnvFile.getCanonicalPath() + " file.")
+            }
+            Properties configuredServerEnvProps = convertServerEnvToProperties(server.serverEnvFile);
+            //Merge with either default server.env or with what has already been merged if
+            mergedProperties = (Properties) combineServerEnvProperties(serverEnvProps, configuredServerEnvProps);
+        }
+
+        if ((server.env != null && !server.env.isEmpty()) || !envProjectProps.isEmpty()) {
+            if (serverEnvPath != null) {
+                logger.warn("The " + serverEnvPath + " file is merged with inlined configuration.")
+            }
```

Comment:
Need to set and return serverEnvPath at the end of this method to indicate what happened in this method. Seeing how this method could do both things (merge serverEnvFile with either the configDir server.env or the default server.env AND merge env props with those as well), I think we should set a flag in each block indicating what got merged and then decide at the end of this method what to set serverEnvPath to. It could be one of the following:

The log message that uses serverEnvPath is as follows:             
logger.info("Update server configuration file server.env from " + serverEnvPath)

1) merging env properties, serverEnvFile server.serverEnvFile.getCanonicalPath(), and configDir server.env serverEnvPath (this is the original one passed in if not null)
2) merging env properties and configDir server.env serverEnvPath (this is the original one passed in if not null)
3) merging serverEnvFile server.serverEnvFile.getCanonicalPath() and configDir server.env serverEnvPath (this is the original one passed in if not null)

Reply:
Should I remove the previous logger.warn messages, or keep them in addition to a logger.info message you suggested?

Reply:
Maybe they should be logger.debug for additional info if they are confused with the result. The logger.info at the end of the main method should be sufficient most of the time.

Reply:
As I was writing this, I realized there are a few more cases to account for. 
4. merging env properties and default server.env and serverEnvFile server.serverEnvFile.getCanonicalPath() (without config dir)

And in all cases, they are also being merged with the default server.env

Reply:
Yes, but "Update server configuration file server.env" is the default server.env.